File: leetcode/0010.Regular-Expression-Matching/RegularExpressionMatching.py
from functools import lru_cache

class Solution:

    # Runtime: 32 ms, faster than 98.87% of Python3 online submissions for Regular Expression Matching.
    def isMatch(self, s: str, p: str) -> bool:
        len_s = len(s)
        len_p = len(p)
        
        @lru_cache(None)
        def dp(i, j):
            if j == len_p:
                return i == len_s
            
            # print("***", i, j, s[i], p[j])
            isFirstOK = i < len_s and p[j] in (".", s[i])

            if j + 1 < len_p and p[j + 1] == '*':
                return dp(i, j + 2) or \
                    isFirstOK and dp(i + 1, j)
            else:
                return isFirstOK and dp(i+1, j+1)
        return dp(0, 0)

if __name__ == "__main__":
    print(Solution().isMatch('aa', "a*"))

File: leetcode/0010.Regular-Expression-Matching/RegularExpressionMatching.go
package leetcode

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Regular Expression Matching.
func dp(i, j int, s, p string, memo map[[2]int]bool) (ans bool) {
	if ans, isSeen := memo[[2]int{i, j}]; isSeen {
		return ans
	}
	if j == len(p) {
		return i == len(s)
	}
	isFirstOK := i < len(s) && (p[j] == '.' || p[j] == s[i])
	if j+1 < len(p) && p[j+1] == '*' {
		ans = (isFirstOK && dp(i+1, j, s, p, memo)) || dp(i, j+2, s, p, memo)
	} else {
		ans = isFirstOK && dp(i+1, j+1, s, p, memo)
	}
	memo[[2]int{i, j}] = ans
	return ans
}

func isMatch(s string, p string) bool {
	memo := make(map[[2]int]bool)
	return dp(0, 0, s, p, memo)
}
